lib/plug/conn.ex

```diff
@@ -1328,50 +1340,149 @@ defmodule Plug.Conn do
+  Once a value is signed or encrypted, you must also call `fetch_cookie/2`
+  with the cookies that are either signed or encrypted.
+
+  To sign, you would do:
+
+      put_resp_cookie(conn, "my-cookie", %{user_id: user.id}, sign: true)
+
+  and then:
+
+      fetch_cookie(conn, signed: ~w(my-cookie))
+
+  To encrypt, you would do:
+
+      put_resp_cookie(conn, "my-cookie", %{user_id: user.id}, encrypt: true)
+
+  and then:
+
+      fetch_cookie(conn, encrypted: ~w(my-cookie))
+
+  By default a signed or encrypted cookie is only valid for a day, unless
+  a `:max_age` is specified.
 
   ## Options
 
     * `:domain` - the domain the cookie applies to
     * `:max_age` - the cookie max-age, in seconds. Providing a value for this
-      option will set both the _max-age_ and _expires_ cookie attributes
+      option will set both the _max-age_ and _expires_ cookie attributes.
     * `:path` - the path the cookie applies to
     * `:http_only` - when `false`, the cookie is accessible beyond HTTP
     * `:secure` - if the cookie must be sent only over https. Defaults
       to true when the connection is HTTPS
     * `:extra` - string to append to cookie. Use this to take advantage of
       non-standard cookie attributes.
+    * `:signed` - when true, signs the cookie
+    * `:encrypted` - when true, encrypts the cookie
 
   """
   @spec put_resp_cookie(t, binary, binary, Keyword.t()) :: t
   def put_resp_cookie(%Conn{} = conn, key, value, opts \\ [])
-      when is_binary(key) and is_binary(value) and is_list(opts) do
+      when is_binary(key) and is_list(opts) do
     %{resp_cookies: resp_cookies, scheme: scheme} = conn
-    cookie = [{:value, value} | opts] |> :maps.from_list() |> maybe_secure_cookie(scheme)
+    {to_send_value, opts} = maybe_sign_or_encrypt_cookie(conn, key, value, opts)
+    cookie = [{:value, to_send_value} | opts] |> Map.new() |> maybe_secure_cookie(scheme)
     resp_cookies = Map.put(resp_cookies, key, cookie)
     update_cookies(%{conn | resp_cookies: resp_cookies}, &Map.put(&1, key, value))
   end
 
+  defp maybe_sign_or_encrypt_cookie(conn, key, value, opts) do
+    {sign?, opts} = Keyword.pop(opts, :sign, false)
+    {encrypt?, opts} = Keyword.pop(opts, :encrypt, false)
+
+    case {sign?, encrypt?} do
+      {true, true} ->
+        raise ArgumentError, "only :sign or :encrypt may be given at the same time"
+
+      {true, false} ->
+        {Plug.Crypto.sign(conn.secret_key_base, key, value, signed_at(opts)), opts}
+
+      {false, true} ->
+        {Plug.Crypto.encrypt(conn.secret_key_base, key, key, value, signed_at(opts)), opts}
+
+      {false, false} when is_binary(value) ->
+        {value, opts}
+
+      {false, false} ->
+        raise ArgumentError, "cookie value must be a binary unless the cookie is signed/encrypted"
+    end
+  end
+
+  defp signed_at(opts) do
+    [
+      keys: Plug.Keys,
+      signed_at: System.system_time(:second) + Keyword.get(opts, :max_age, 86400)
```

Comment:
Yes, so the trick here is to move `signed_at` to the future and set a `max_age` of zero since, in cookies, max-age is given when writing and not on reading. I will add some code comments explaining this is the case to avoid further confusion.

Reply:
Just beware that things will break if you ever decide to check in Plug.Crypto that `signed_at` is not in the past. Which might be a good thing to do, generally, since leaving it open ended on one end gives an attacker quite a lot more wriggle room than requiring the timestamp to be in a specific window of [signed_at ... signed_at + max_age].

Reply:
That’s a good point. I will provide proper support for this on Plug.Crypto.

Reply:
I know allow setting `max_age` when signing/encrypting. Thanks for keeping me honest!